sequencer/src/block2.rs

```diff
@@ -0,0 +1,144 @@
+use std::mem::size_of;
+
+#[allow(dead_code)] // TODO temporary
+pub struct BlockPayload {
+    payload: Vec<u8>,
+    // tx_ranges: Vec<Range<u32>>, // not convinced we need this
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Tx {
+    // namespace: u32, // TODO how many bytes for namespace id?
+    payload: Vec<u8>,
+}
+
+impl BlockPayload {
+    #[allow(dead_code)] // TODO temporary
+    fn build(txs: impl IntoIterator<Item = Tx>) -> Self {
+        // tx_table[i] is the end index (exclusive) of the ith tx
+        // so that the payload bytes of the ith tx is
+        // tx_bodies[tx_table[i-1]..tx_table[i]].
+        // edge case: tx_table[-1] is defined as 0.
+        //
+        // TODO clarification for future docs:
+        // perhaps the final entry of tx_table is redundant:
+        // it should always equal the length of the flattened tx bytes.
+        // But this might not always be the case.
+        // Example: If the flattened tx bytes are not a monolith at the end
+        // of the payload then the final tx_table entry is needed.
+        let mut tx_table = Vec::new();
+
+        // concatenation of all tx payloads
+        let mut tx_bodies = Vec::new();
+
+        // build tx_table, tx_bodies
+        let mut end: u32 = 0;
+        for tx in txs.into_iter() {
+            // TODO idiomatic usize -> u32 conversion?
+            // Is panic acceptable? If not then we need to return `Result`.
+            let len: u32 = tx
+                .payload
+                .len()
+                .try_into()
+                .expect("tx byte length should fit into u32");
+
+            end = end
+                .checked_add(len)
+                .expect("total byte length of all tx bodies should fit into u32");
+
+            // TODO to_be_bytes or to_le_bytes?
+            tx_table.extend(end.to_be_bytes());
+            tx_bodies.extend(tx.payload);
+        }
+
+        // tx_table_len is the number of 4-byte entries.
+        // tx_table.len() is the number of bytes in the tx table,
+        // so we divide by size_of::<u32>()
+        let tx_table_len: u32 = (tx_table.len() / size_of::<u32>())
+            .try_into()
+            .expect("tx_table len should fit into u32");
+
+        // Naively copy all pieces into a flat payload.
+        // We could avoid this by allocating memory in advance,
+        // but that would require knowing the number of txs and their total
+        // byte length in advance, which we can't do without a complete scan
+        // of the `txs` iterator.
+        let mut payload = Vec::new();
+        payload.extend(tx_table_len.to_be_bytes());
+        payload.extend(tx_table);
+        payload.extend(tx_bodies);
+        Self { payload }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{BlockPayload, Tx};
+    use std::mem::size_of;
+
+    #[test]
+    fn build_basic_correctness() {
+        // play with this
+        let test_cases = vec![
```

Comment:
These are good test cases. Just a note for the future, when we get around to functions that read the payload, like transaction getters, we'll need to test with some blocks that weren't built by `BlockPayload::build`, since that function only builds "nice" blocks. We'll need to maliciously build some blocks with overlapping and truncated and out-of-order transactions.

Reply:
Roger that. At a bare minimum we'll need those tests for the implementation of `decode` / `from_bytes`.